src/rust/engine/src/nodes.rs

```diff
@@ -515,43 +562,33 @@ pub struct Task {
 }
 
 impl Step for Task {
-  fn step(&self, context: StepContext) -> State<Node> {
-    // Compute dependencies for the Node, or determine whether it is a Noop.
-    let mut dependencies = Vec::new();
-    let mut dep_values: Vec<&Value> = Vec::new();
-    for selector in &self.selector.clause {
-      let dep_node =
-        Node::create(
-          selector.clone(),
-          self.subject.clone(),
-          self.variants.clone()
-        );
-      match context.get(&dep_node) {
-        Some(&Complete::Return(ref value)) =>
-          dep_values.push(&value),
-        Some(&Complete::Noop(_, _)) =>
-          return State::Complete(
-            Complete::Noop("Was missing (at least) input {}.", Some(dep_node))
-          ),
-        Some(&Complete::Throw(ref msg)) =>
-          // NB: propagate thrown exception directly.
-          return State::Complete(Complete::Throw(context.clone_val(msg))),
-        None =>
-          dependencies.push(dep_node),
-      }
-    }
+  fn step(&self, context: Context) -> StepFuture {
+    let deps =
+      future::join_all(
+        self.selector.clause.iter()
+          .map(|selector| {
+            context.get(&Node::create(selector.clone(), self.subject.clone(), self.variants.clone()))
+          })
+          .collect::<Vec<_>>()
+      );
 
-    if !dependencies.is_empty() {
-      // A clause was still waiting on dependencies.
-      State::Waiting(dependencies)
-    } else {
-      // Ready to run!
-      State::Runnable(Runnable {
-        func: self.selector.func,
-        args: dep_values.iter().map(|v| context.clone_val(v)).collect(),
-        cacheable: self.selector.cacheable,
+    let selector = self.selector.clone();
+    deps
+      .then(move |deps_result| {
+        match deps_result {
+          Ok(deps) =>
+            context.invoke_runnable(
+              Runnable {
+                func: selector.func,
+                args: deps.iter().map(|v| context.clone_val(v)).collect(),
+                cacheable: selector.cacheable,
+              }
+            ),
+          Err(err) =>
+            Err(context.was_optional(err, "Missing at least one input.")),
+        }
       })
-    }
+      .boxed()
```

Comment:
was reading through this syntax and came across https://github.com/alexcrichton/futures-rs/issues/228